### ISSUE TITLE
Detailed job info for completed `CalcJobs` gets set to `None`

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#3133]

**Comment by giovannipizzi** [comment]
I think this is very relevant info that we should try to put back soon.
The logic from the Scheduler plugin shouldn't have changed - simply for some reason the value gets now set to None by the engine.

**Comment by sphuber** [comment]
The reason this fails is that currently the code that is supposed to do it is in the wrong location I feel. It is currently placed within the `JobsList` construct, whose purpose it is to bundle all job status update requests, do it once for all jobs and then update the individual jobs. The current logic checks on the current state of the job returned by the scheduler and if it is `DONE` tries to get the detailed information. Note that this will make an individual call to the scheduler (`sacct` for SLURM) for each job that is managed by the `JobsList`. The bigger problem however, is that `DONE` will almost never be hit as when the job is done it won't be in the response.

What seems more logical to me is to add another transport task. Instead of going straight to `RETRIEVE` after `UPDATE` is done, instead go to `ACCOUNTING`, which will try to call `get_detailed_job_info` and will parse it. The "downside" is that it takes a separate transport task but there are many benefits. It is couched in its own exponential backoff loop and is not muffled in with another task. Plus logically, this makes a lot more sense. The current way of in each update cycle for all jobs, making an additional call for those that are finished (which currently is not even working) is fudging in a lot of additional scheduler operations and is mixing responsibilities.

I also have a question what the reasoning was behind storing the `detailed_jobinfo` within the `last_jobinfo` attribute. Doesn't it make more sense to have it in parallel or even have it replace the `last_jobinfo`. In principle the `detailed_jobinfo` should have all the properties and more that would have been in the `last_jobinfo`. This way there is no overlap. I am not sure what the added value is to have partial information of the last time the scheduler was queried when the job was still running.

So my suggestion is the following. Add a new transport task `ACCOUNTING` between `UPDATE` and `RETRIEVE` that will fetch detailed job info, parse it (see PR #3261 by @espenfl ) into a dictionary and store it as `detailed_job_info` attribute

**Comment by giovannipizzi** [comment]
> So my suggestion is the following. Add a new transport task ACCOUNTING between UPDATE and RETRIEVE that will fetch detailed job info, parse it (see PR #3261 by @espenfl ) into a dictionary and store it as detailed_job_info attribute

I like the way you think! ;-) Ok with me!

> what the reasoning was behind storing the detailed_jobinfo within the last_jobinfo attribute

Not sure I'm going to reply to your question, but you don't want - I think - to ask the detailed information at every iteration; you ask them only at the end. And the reason is mainly for having it for debugging later, and we do it because this information is often deleted by the scheduler after a few days.
It's the 'last' job info in the sense that it's the done at the end.
If we want to change names to make them more consistent it's ok with me (e.g. calling it `detailed_job_info`, and documenting when it is retrieved).

Thanks!

**Comment by espenfl** [comment]
Aha, so this was in fact an old issue. Okey, see my comments in #3280 as well.

Yes, adding an accounting step makes total sense. I support this. I can have a look in this in the PR and suggest such a change where I also remove the `detailed_job_info` pull during `UPDATE` and only do it in an `ACCOUNTING` step. Also, I think it makes sense to follow the lines in the comment in #3280. Injecting new transport tasks is entirely new to me. Happy to learn, but some pointers on how to approach this would be great. @sphuber understand you leave for vacation, so happy to look at this while you are away.

**Comment by sphuber** [comment]
I have already made the implementation. I am cleaning it up now a bit and will then push the branch to my fork. It implements the basic `ACCOUNTING` infrastructure and then adds the SLURM implementation, based on the work in your PR. This will already then put a parsed dictionary in a node attribute. We can then build on top of this to use this information, together with scheduler stdout and stderr to parse known error modes, e.g. OOM and OOW and set those as exit statuses before we call the actual parser, if any. The developer of the parser can then inspect this exit status and try to parse any of the other retrieved files and decide to override with something else if applicable.

**Comment by espenfl** [comment]
Great @sphuber. Will have a look at this tomorrow.

**Comment by giovannipizzi** [comment]
Some updates after discussing this with @sphuber 

- while the idea of having a different step is formally nice, we realised that this will on average make all calculation finish much later than needed (if there is only 1 calculation and it goes over SSH, there will be ~30s to wait just for the accounting, and this seems a big cost). Therefore we will pack the logic at the beginning of the retrieval step, in the same transport context manager (if this fails, anyway the whole retrieval step will be retried; if the file retrieval fails, it shouldn't be too bad to retrieve again the detailed job info).
- in terms of the parsing of the detailed job info, and homogenising the values (see `_MAP_JOB_ACCOUNTING_SLURM` as suggested by @espenfl in his PR): I personally feel that it makes sense to parse and convert the names to more human-readable names only if we can homogenise this among schedulers. However, this is quite challenging task (info is quite diverse, and also we don't have access to the output of many of the non-SLURM schedulers). My intention is therefore to keep this as a 'log file', and not parse it nor try to homogenise. Maybe the only discussion to have is if we want to keep this as a string, or parse it into a list of dictionaries without renaming the fields. The latter is useful only if we need to query. But I feel that this is something that needs to be done only by few people, so *I would vote for*: keeping it a string, and add some parsing utility function in the code since Espen already wrote them; and possibly documenting (e.g. in the cookbook) how to call it recursively on N nodes and add the parsed results to the extras, for later parsing). We can discuss this tonight or tomorrow with @espenfl when he arrives.

Some technical tasks that need to be taken care of, on top of @sphuber's branch:
- [x] fix some typos `SCHEUDLER`->`SCHEDULER`
- [ ] Add flag in authinfo to skip accounting if you know you don't care
- [ ] Make it possible to list the fact the accounting function is not implemented (`raise NotImplementedError`) among the scheduler `features`
- [ ] Add a `get_dict()` and a `load_from_dict()` in `scheduler.dataclasses.JobInfo` instead of only a serialiser, so we can store in the DB a dictionary and not a JSON string for the `last_jobinfo` in [this part of the code](https://github.com/aiidateam/aiida-core/blob/0644e7d9ff492fdb2c94d608c575c697daee6d32/aiida/orm/nodes/process/calculation/calcjob.py#L423), i.e. the cache of the latest time the engine saw the job (we will do this without DB migrations).
